### PR TITLE
RF01 & recursive_crawl improvements

### PR DESCRIPTION
**Pull request by alanmcruickshank**
This is a fairly small PR, aimed at improving performance of RF01.

Includes one syntax simplification in RF01 (which is trivial), but more importantly updating the `.recursive_crawl()` method to filter based on the same type sets we now use in rule crawling.

Early testing suggests this more than halves the time spent in RF01 in long queries. Across a blended project the impact is more like 10%. Similar gains on ST05.

### CODE REVIEW

File: src/sqlfluff/rules/references/RF01.py [hunk 1]
```diff
@@ -91,30 +91,30 @@ def _eval(self, context: RuleContext) -> EvalResultType:
         ):
             return LintResult()
 
+        if FunctionalContext(context).parent_stack.any(sp.is_type(*_START_TYPES)):
```
Comment (pwildenhain) [inline src/sqlfluff/rules/references/RF01.py:94]: Good refactor here

File: src/sqlfluff/rules/structure/ST05.py [hunk 2]
```diff
@@ -531,13 +531,21 @@ def _create_table_ref(table_name: str, dialect: Dialect) -> TableExpressionSegme
 
 
 def _get_case_preference(root_select: Segments):
-    first_keyword = root_select.recursive_crawl(
-        "keyword",
-        recurse_into=False,
-    ).first()[0]
-    if first_keyword.raw[0].islower():
```
Comment (pwildenhain) [inline src/sqlfluff/rules/structure/ST05.py:538]: I'm sure mypy also appreciates this refactor too

**Review by pwildenhain — APPROVED** [review]
🚀

File: src/sqlfluff/core/parser/segments/base.py [hunk 3]
```diff
@@ -1145,17 +1145,29 @@ def recursive_crawl(
...
         if self.is_type(*seg_type):
             match = True
             yield self
         else:
             match = False
+
+        # Check whether the types we're looking for are in this segment
+        # at all. If not, exit early.
+        if not self.descendant_type_set.intersection(seg_type):
+            return
```
Comment (WittierDinosaur) [inline src/sqlfluff/core/parser/segments/base.py:1162]:
```suggestion
            return None
```

File: src/sqlfluff/rules/references/RF01.py [hunk 4]
```diff
@@ -91,30 +91,30 @@ def _eval(self, context: RuleContext) -> EvalResultType:
...
+            return LintResult()
```
Comment (WittierDinosaur) [inline src/sqlfluff/rules/references/RF01.py:95]: My only issue with this is the mypy strict stuff. Between this and the below, we have 3 types of return, List[LintResult] (violations), LintResult, or None - could do with a bit of cleaning. Really we should only have one Falsy return type

**Review by WittierDinosaur — CHANGES_REQUESTED** [review]
Two small things to keep us honest with mypy

File: src/sqlfluff/rules/references/RF01.py (hunk 4 above)
Comment (alanmcruickshank) [reply src/sqlfluff/rules/references/RF01.py]: That's a good point - I've made them all lists, which I think is best in this case

**Review by WittierDinosaur — APPROVED** [review]
LGTM

**Pull request merged**